consensus/hotstuff/pacemaker/pacemaker.go

```diff
@@ -12,111 +12,152 @@ import (
+	return p.livenessData.CurrentView
 }
 
 // TimeoutChannel returns the timeout channel for current active timeout.
 // Note the returned timeout channel returns only one timeout, which is the current
 // timeout.
 // To get the timeout for the next timeout, you need to call TimeoutChannel() again.
-func (p *NitroPaceMaker) TimeoutChannel() <-chan time.Time {
+func (p *ActivePaceMaker) TimeoutChannel() <-chan time.Time {
 	return p.timeoutControl.Channel()
 }
 
 // ProcessQC notifies the pacemaker with a new QC, which might allow pacemaker to
-// fast forward its view.
-func (p *NitroPaceMaker) ProcessQC(qc *flow.QuorumCertificate) (*model.NewViewEvent, bool) {
-	if qc.View < p.currentView {
-		return nil, false
+// fast-forward its view.
+func (p *ActivePaceMaker) ProcessQC(qc *flow.QuorumCertificate) (*model.NewViewEvent, error) {
+	if qc.View < p.CurView() {
+		return nil, nil
 	}
+
+	p.timeoutControl.OnProgressBeforeTimeout()
+
 	// qc.view = p.currentView + k for k ≥ 0
 	// 2/3 of replicas have already voted for round p.currentView + k, hence proceeded past currentView
 	// => 2/3 of replicas are at least in view qc.view + 1.
 	// => replica can skip ahead to view qc.view + 1
-	p.timeoutControl.OnProgressBeforeTimeout()
-
 	newView := qc.View + 1
+	err := p.updateLivenessData(newView, qc, nil)
+	if err != nil {
+		return nil, err
+	}
+
 	p.notifier.OnQcTriggeredViewChange(qc, newView)
-	return p.gotoView(newView), true
+	return &model.NewViewEvent{View: newView}, nil
 }
 
-func (p *NitroPaceMaker) ProcessTC(tc *flow.TimeoutCertificate) (*model.NewViewEvent, bool) {
-	panic("not yet implemented")
+func (p *ActivePaceMaker) ProcessTC(tc *flow.TimeoutCertificate) (*model.NewViewEvent, error) {
+	if tc == nil || tc.View < p.CurView() {
+		return nil, nil
+	}
+
+	newView := tc.View + 1
+	err := p.updateLivenessData(newView, tc.NewestQC, tc)
+	if err != nil {
+		return nil, err
+	}
+
+	p.notifier.OnTcTriggeredViewChange(tc, newView)
+	return &model.NewViewEvent{View: newView}, nil
 }
 
-func (p *NitroPaceMaker) OnPartialTC(curView uint64) {
-	panic("not yet implemented")
+func (p *ActivePaceMaker) OnPartialTC(newView uint64) {
+	if p.CurView() == newView {
+		p.timeoutControl.TriggerTimeout()
+	}
 }
 
 // NewestQC returns QC with the highest view discovered by PaceMaker.
-func (p *NitroPaceMaker) NewestQC() *flow.QuorumCertificate {
-	panic("not yet implemented")
+func (p *ActivePaceMaker) NewestQC() *flow.QuorumCertificate {
+	return p.livenessData.NewestQC
 }
 
 // LastViewTC returns TC for last view, this could be nil if previous round
 // has entered with a QC.
-func (p *NitroPaceMaker) LastViewTC() *flow.TimeoutCertificate {
-	panic("not yet implemented")
+func (p *ActivePaceMaker) LastViewTC() *flow.TimeoutCertificate {
+	return p.livenessData.LastViewTC
 }
 
 // Start starts the pacemaker
-func (p *NitroPaceMaker) Start() {
+func (p *ActivePaceMaker) Start() {
 	if p.started.Swap(true) {
 		return
 	}
```

Comment:
This purpose of `started` is to ensure we only start the initial timer once. To me it would be clearer to use `sync.Once` for this instead, since it's made exactly for that purpose.

Reply:
+1